transporters/src/main/java/com/nextdoor/bender/ipc/s3/S3TransportFactory.java

```diff
@@ -55,32 +54,33 @@ public Transport newInstance() throws TransportFactoryInitException {
 
   @Override
   public void close() {
-    Exception e = null;
     for (MultiPartUpload upload : this.pendingMultiPartUploads.values()) {
-
-      if (e == null) {
-        CompleteMultipartUploadRequest req = upload.getCompleteMultipartUploadRequest();
-        try {
-          this.client.completeMultipartUpload(req);
-        } catch (AmazonS3Exception e1) {
-          logger.error("failed to complete multi-part upload for " + upload.getKey() + " parts "
-              + upload.getPartCount(), e1);
-          e = e1;
-        }
-      } else {
-        logger.warn("aborting upload for: " + upload.getKey());
-        AbortMultipartUploadRequest req = upload.getAbortMultipartUploadRequest();
-        try {
-          this.client.abortMultipartUpload(req);
-        } catch (AmazonS3Exception e1) {
-          logger.error("failed to abort multi-part upload", e1);
-        }
+      CompleteMultipartUploadRequest req = upload.getCompleteMultipartUploadRequest();
+      try {
+        this.client.completeMultipartUpload(req);
+      } catch (AmazonS3Exception e) {
+        String errorMessage = String.format("failed to complete multi-part upload for key %s, uploadId %s, and partCount %s.",
+                upload.getKey(),
+                upload.getPartCount(),
+                upload.getUploadId());
+        logger.error(errorMessage, e);
+        abortAllMultiPartUploads();
+        throw new RuntimeException(errorMessage + " . All remaining uploads were aborted as a result.", e);
```

Comment:
You need to throw the exception outside of the loop because otherwise we won't iterate over the remaining uploads that need to be canceled. Additionally we won't get to `this.pendingMultiPartUploads.clear();` which is important.

Reply:
The abortAllMultiPartUploads() handles aborting all the uploads in the map automatically instead of storing the error object and checking each time. 

I can add the `this.pendingMultiPartUploads.clear();` to the catch block as well. I searched around for how the map of uploads was used after the flush but didn't see occurrences though.

Reply:
Ah, it shouldn't be used but for safety it should be cleared. Also, I prefer derefrencing explicitly when being done with an object.

Reply:
sure no worries.